File: client/lib/src/models/device_info.dart
part of form_concierge_client;

class DeviceInfo {
  final String? deviceId;
  final String? label;
  final String? platform;
  final String? os;
  final String? osVersion;
  final String? browser;
  final String? browserVersion;
  final String? appVersion;
  final String? appBuild;
  final String? model;
  final String? manufacturer;
  final String? locale;
  final String? timezone;
  final int? screenWidth;
  final int? screenHeight;
  final double? devicePixelRatio;
  final String? userAgent;

  const DeviceInfo({
    this.deviceId,
    this.label,
    this.platform,
    this.os,
    this.osVersion,
    this.browser,
    this.browserVersion,
    this.appVersion,
    this.appBuild,
    this.model,
    this.manufacturer,
    this.locale,
    this.timezone,
    this.screenWidth,
    this.screenHeight,
    this.devicePixelRatio,
    this.userAgent,
  });

  factory DeviceInfo.fromJson(Map<String, dynamic> json) => DeviceInfo(
    deviceId: _optionalString(json['deviceId']),
    label: _optionalString(json['label']),
    platform: _optionalString(json['platform']),
    os: _optionalString(json['os']),
    osVersion: _optionalString(json['osVersion']),
    browser: _optionalString(json['browser']),
    browserVersion: _optionalString(json['browserVersion']),
    appVersion: _optionalString(json['appVersion']),
    appBuild: _optionalString(json['appBuild']),
    model: _optionalString(json['model']),
    manufacturer: _optionalString(json['manufacturer']),
    locale: _optionalString(json['locale']),
    timezone: _optionalString(json['timezone']),
    screenWidth: json['screenWidth'] == null ? null : _int(json['screenWidth']),
    screenHeight: json['screenHeight'] == null
        ? null
        : _int(json['screenHeight']),
    devicePixelRatio: json['devicePixelRatio'] == null
        ? null
        : _double(json['devicePixelRatio']),
    userAgent: _optionalString(json['userAgent']),
  );

  Map<String, dynamic> toJson() => _withoutNulls({
    'deviceId': deviceId,
    'label': label,
    'platform': platform,
    'os': os,
    'osVersion': osVersion,
    'browser': browser,
    'browserVersion': browserVersion,
    'appVersion': appVersion,
    'appBuild': appBuild,
    'model': model,
    'manufacturer': manufacturer,
    'locale': locale,
    'timezone': timezone,
    'screenWidth': screenWidth,
    'screenHeight': screenHeight,
    'devicePixelRatio': devicePixelRatio,
    'userAgent': userAgent,
  });

  DeviceInfo merge(DeviceInfo? override) {
    if (override == null) return this;
    return DeviceInfo(
      deviceId: override.deviceId ?? deviceId,
      label: override.label ?? label,
      platform: override.platform ?? platform,
      os: override.os ?? os,
      osVersion: override.osVersion ?? osVersion,
      browser: override.browser ?? browser,
      browserVersion: override.browserVersion ?? browserVersion,
      appVersion: override.appVersion ?? appVersion,
      appBuild: override.appBuild ?? appBuild,
      model: override.model ?? model,
      manufacturer: override.manufacturer ?? manufacturer,
      locale: override.locale ?? locale,
      timezone: override.timezone ?? timezone,
      screenWidth: override.screenWidth ?? screenWidth,
      screenHeight: override.screenHeight ?? screenHeight,
      devicePixelRatio: override.devicePixelRatio ?? devicePixelRatio,
      userAgent: override.userAgent ?? userAgent,
    );
  }

  String? get summary {
    final name = label ?? model;
    final system = [
      os,
      osVersion,
    ].whereType<String>().where((value) => value.isNotEmpty).join(' ');
    final agent = [
      browser,
      browserVersion,
    ].whereType<String>().where((value) => value.isNotEmpty).join(' ');
    final parts = [
      name,
      system.isEmpty ? platform : system,
      agent.isEmpty ? null : agent,
    ].whereType<String>().where((value) => value.isNotEmpty).toList();
    return parts.isEmpty ? null : parts.join(' / ');
  }

  String? get detailSummary {
    final screen = screenWidth != null && screenHeight != null
        ? '${screenWidth}x$screenHeight'
        : null;
    final parts = [
      locale,
      timezone,
      screen,
      devicePixelRatio == null ? null : '${devicePixelRatio}x',
      appVersion == null ? null : 'app $appVersion',
      appBuild == null ? null : 'build $appBuild',
    ].whereType<String>().where((value) => value.isNotEmpty).toList();
    return parts.isEmpty ? null : parts.join(' / ');
  }
}

File: swiftui/Sources/FormConciergeSwiftUI/DeviceInfo.swift
import Foundation

#if canImport(UIKit)
  import UIKit
#endif
#if canImport(AppKit)
  import AppKit
#endif

public struct DeviceInfo: Codable, Sendable {
  public let deviceId: String?
  public let label: String?
  public let platform: String?
  public let os: String?
  public let osVersion: String?
  public let browser: String?
  public let browserVersion: String?
  public let appVersion: String?
  public let appBuild: String?
  public let model: String?
  public let manufacturer: String?
  public let locale: String?
  public let timezone: String?
  public let screenWidth: Int?
  public let screenHeight: Int?
  public let devicePixelRatio: Double?
  public let userAgent: String?

  public init(
    deviceId: String? = nil,
    label: String? = nil,
    platform: String? = nil,
    os: String? = nil,
    osVersion: String? = nil,
    browser: String? = nil,
    browserVersion: String? = nil,
    appVersion: String? = nil,
    appBuild: String? = nil,
    model: String? = nil,
    manufacturer: String? = nil,
    locale: String? = nil,
    timezone: String? = nil,
    screenWidth: Int? = nil,
    screenHeight: Int? = nil,
    devicePixelRatio: Double? = nil,
    userAgent: String? = nil
  ) {
    self.deviceId = deviceId
    self.label = label
    self.platform = platform
    self.os = os
    self.osVersion = osVersion
    self.browser = browser
    self.browserVersion = browserVersion
    self.appVersion = appVersion
    self.appBuild = appBuild
    self.model = model
    self.manufacturer = manufacturer
    self.locale = locale
    self.timezone = timezone
    self.screenWidth = screenWidth
    self.screenHeight = screenHeight
    self.devicePixelRatio = devicePixelRatio
    self.userAgent = userAgent
  }

  public static var current: DeviceInfo {
    var label: String?
    var platform = "swiftui"
    var os: String?
    var osVersion = ProcessInfo.processInfo.operatingSystemVersionString
    var screenWidth: Int?
    var screenHeight: Int?
    var scale: Double?

    #if canImport(UIKit)
      let device = UIDevice.current
      label = device.model
      platform = "ios"
      os = device.systemName
      osVersion = device.systemVersion
      let size = UIScreen.main.bounds.size
      screenWidth = Int(size.width.rounded())
      screenHeight = Int(size.height.rounded())
      scale = UIScreen.main.scale
    #elseif canImport(AppKit)
      label = Host.current().localizedName
      platform = "macos"
      os = "macOS"
      if let frame = NSScreen.main?.frame {
        screenWidth = Int(frame.width.rounded())
        screenHeight = Int(frame.height.rounded())
      }
      scale = NSScreen.main.map { Double($0.backingScaleFactor) }
    #endif

    return DeviceInfo(
      label: label,
      platform: platform,
      os: os,
      osVersion: osVersion,
      locale: Locale.current.identifier,
      timezone: TimeZone.current.identifier,
      screenWidth: screenWidth,
      screenHeight: screenHeight,
      devicePixelRatio: scale
    )
  }
}
